File: 1143-longest-common-subsequence/1143-longest-common-subsequence.cpp
class Solution {
public:
int dp[1001][1001];
int rec(int i,int j,string&text1,string&text2){
    if(i>=text1.size() || j>=text2.size())
    {
        return 0;
    }

    if(dp[i][j]!=-1)return dp[i][j];

    int cnt=0;
    if(text1[i]==text2[j])
    {
        cnt=max(cnt,rec(i+1,j+1,text1,text2)+1);
    }
    cnt=max(cnt,rec(i+1,j,text1,text2));
    cnt=max(cnt,rec(i,j+1,text1,text2));
    return dp[i][j]=cnt;
}
    int longestCommonSubsequence(string text1, string text2) {
        memset(dp,-1,sizeof(dp));
        return rec(0,0,text1,text2);
    }
};

File: 1143-longest-common-subsequence/1143-longest-common-subsequence.py
class Solution:

    def rec(self,i,j,text1,text2,dp):
        if i>=len(text1) or j>=len(text2):
            return 0
        if self.dp[i][j]!=-1:
            return self.dp[i][j]
        cnt=0
        if text1[i]==text2[j]:
            cnt=max(cnt,self.rec(i+1,j+1,text1,text2,dp)+1)
        cnt=max(cnt,self.rec(i+1,j,text1,text2,dp))
        cnt=max(cnt,self.rec(i,j+1,text1,text2,dp))

        self.dp[i][j]=cnt
        return cnt

    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        self.dp=[[-1] * 1001 for _ in range(1001)]
        return self.rec(0,0,text1,text2,self.dp)

        